### PR TITLE
Multi decompiler + FernFlower

### PR DESCRIPTION
**Pull request by ielali**
Hello again, finally had some time to work on this project.

I ended up creating a branch with a drop down for all the compilers I've discovered so far. Even if we don't go with that option it's still very useful to compare and choose the decompiler for the next release.

I was finally able to make FurnFlower work but not without difficulty. As I mentionned before there are no official maven releases for this decompiler since it's destined to be an IntelliJ plugin. The only mirror that does a release is 4 years old, quite stale. So I ended up including the jar in the project and adding a maven system dependency. The maven bundle plugin needed an upgrade for this type of dependency to be embedded correctly. I also had to implement code to be able to target a single class in the bundle jar.

I didn't find any improvements over JD-Core which in my opinion still produce the best source code output. But I will let you test on your own and come back to me.

### CODE REVIEW

**Comment by pun-ky** [comment]
great job. I was suspecting that Fernflower will produce the best outputs. So interesting. I will check that JD too.

Now I need to find some time to work on this too. It will be not so easy but I hope during next week I will prepare feedback for you. And I am not saying no to the multi-compiler release. If some compiler will not work on some runtime, there will be an option to try another one.

About embedding Fernflower and system dependency. Don't worry.  I don't take care. It's rather dev-tool ;) It just must work, that's all ;)

File: src/main/resources/search/scripts.js [hunk 1]
```diff
@@ -83,13 +83,20 @@ $(function () {
 
     // Viewer for decompiled class source
     var classDecompileTemplate = Handlebars.compile($('#class-decompile-template').html());
+
     function classDecompile(url) {
+        let params = {
+            "decompiler": $("select[name='decompiler']", $form).val(),
+            "lineNumbers": $("select[name='lineNumbers']", $form).val()
```
Comment (pun-ky) [inline src/main/resources/search/scripts.js:90]: what does that option 'lineNumbers' do?

File: src/main/resources/search/plugin.html [hunk 2]
```diff
@@ -54,6 +54,21 @@
...
 				</label>
+				<label title="Choose the Java decompiler library to use">
+					<span>Decompiler</span>
+					<select name="decompiler" class="ui-state-default ui-corner-all">
+						<option value="JD_CORE" selected>JD-Core</option>
+						<option value="FERN_FLOWER">FernFlower</option>
+						<option value="CFR">CFR</option>
+					</select>
+				</label>
+				<label title="Whether to show/hide line number in decompiled classes when supported">
```
Comment (pun-ky) [inline src/main/resources/search/plugin.html:65]: cool! ;)

**Review by pun-ky — APPROVED** [review]
The code looks totally ok, even changes in UI also look good.
I am excited before compiling and testing it on my own.

File: pom.xml [hunk 3]
```diff
@@ -207,23 +232,22 @@
...
-
         <dependency>
-            <groupId>org.bitbucket.mstrobel</groupId>
-            <artifactId>procyon-compilertools</artifactId>
-            <version>${procyon.version}</version>
+            <groupId>org.jetbrains.java.decompiler</groupId>
+            <artifactId>fernflower</artifactId>
+            <version>1.0</version>
+            <scope>system</scope>
+            <systemPath>${project.basedir}/jars/fernflower-decompiler.jar</systemPath>
```
Comment (bartoszWesolowski) [inline pom.xml]: Should it be `jar-dependencies` ?

Comment (ielali) [reply pom.xml]: Nope pick any name you'd like for me naming is the most difficult part :)

Comment (ielali) [reply pom.xml]: Sorry just fixed it didn't notice it was not renamed in the pom. Thought you were providing a different name suggestion.

**Comment by ielali** [comment]
One more proposal, bundle tree exploration source displayed in the same window for an easier exploration within a given bundle. I used a simple grid css based layout. Let me know what you think.

**Comment by pun-ky** [comment]
just cloned your branch and tried building... :/

File: pom.xml [hunk 4]
```diff
@@ -272,6 +296,7 @@
         <felix.url>http://localhost:8080/system/console</felix.url>
         <felix.user>admin</felix.user>
         <felix.password>admin</felix.password>
-        <procyon.version>0.5.36</procyon.version>
+        <jd-core.version>1.1.3</jd-core.version>
+        <cfr.version>0.152-SNAPSHOT</cfr.version>
```
Comment (pun-ky) [inline pom.xml]: why snapshot?

**Review by pun-ky — CHANGES_REQUESTED** [review]

**Comment by pun-ky** [comment]
I corrected cfr to 0.151 and now

**Comment by pun-ky** [comment]
ok, switched java from 8 to 11 and works.... this is new constraint unfortunately :/

File: pom.xml (hunk 4 above)
Comment (ielali) [reply pom.xml]: Yes sorry about that initially I had the car project checked out locally. Had to experiment to make it decompile a single class

**Comment by pun-ky** [comment]
how about that ? "could not load..."

**Comment by pun-ky** [comment]
CFR seems to not work properly, inner classes not working :/ other compilers are working well with that class:

**Comment by ielali** [comment]
With fernflower I had this issue with inner classes and I was forced to extract to a temporary directory all inner classes before being able to decompile. Maybe need the same fix.

**Comment by pun-ky** [comment]
css need to be fixed; finally, I could correct styling, no worries ;)

**Comment by pun-ky** [comment]
also does not work on Sling Starter 11 ; IMO it is must have; previous version was working even on pure Felix / pure OSGi

**Comment by pun-ky** [comment]
3 decompilers with that switch is generally fancy and I am opting for keeping it. However, it depends on you and your power to provide enough stability/quality in the final implementation.

saying that the next version will be 2.0.0 and java supported will be at least 11 is also fine.

so keep up at good work ;)

finally, I will probably tune a little bit UI then release it but at first, please resolve CFR issues

**Comment by ielali** [comment]
CFR inner class issue corrected, working fine in AEM (6.5).  Found the issue with Sling Starter, the bundle jar resolution was not working. Apparently it's now sourcing jars directly from maven repository. So now if the jar is not found through the bundle dir we try to resolve the bundle location. Don't know if you have a better approach.
`| ~/Development/Tools/sling-org-apache-sling-starter/launcher/framework @ C02WP09MHV2R (imad.elali)  (master) 
| =>  cat bundle99/bundle.info 
99
reference:file:/Users/imad.elali/.m2/repository/org/apache/jackrabbit/oak-jackrabbit-api/1.32.0/oak-jackrabbit-api-1.32.0.jar
32
15
1614898569593
0`

**Comment by pun-ky** [comment]
@ielali I remember about work you did and I will consume it when I find some time for sure ;)

**Comment by pun-ky** [comment]
hmm strange

```
java.lang.IllegalArgumentException: No enum constant com.neva.felix.webconsole.plugins.search.decompiler.Decompilers.
	at java.base/java.lang.Enum.valueOf(Enum.java:240)
	at com.neva.felix.webconsole.plugins.search.decompiler.Decompilers.valueOf(Decompilers.java:3)
	at com.neva.felix.webconsole.plugins.search.rest.ClassDecompileServlet.doGet(ClassDecompileServlet.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.apache.felix.http.base.internal.handler.ServletHandler.handle(ServletHandler.java:123)
	at org.apache.felix.http.base.internal.dispatch.InvocationChain.doFilter(InvocationChain.java:86)
	at com.adobe.granite.license.impl.LicenseCheckFilter.doFilter(LicenseCheckFilter.java:308)

```

**Comment by ielali** [comment]
> hmm strange
> 
> ```
> java.lang.IllegalArgumentException: No enum constant com.neva.felix.webconsole.plugins.search.decompiler.Decompilers.
> 	at java.base/java.lang.Enum.valueOf(Enum.java:240)
> 	at com.neva.felix.webconsole.plugins.search.decompiler.Decompilers.valueOf(Decompilers.java:3)
> 	at com.neva.felix.webconsole.plugins.search.rest.ClassDecompileServlet.doGet(ClassDecompileServlet.java:71)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
> 	at org.apache.felix.http.base.internal.handler.ServletHandler.handle(ServletHandler.java:123)
> 	at org.apache.felix.http.base.internal.dispatch.InvocationChain.doFilter(InvocationChain.java:86)
> 	at com.adobe.granite.license.impl.LicenseCheckFilter.doFilter(LicenseCheckFilter.java:308)
> ```

In with environment did you get this? the bundle is working fine in AEM 6.5

**Comment by krystian-panek-vmltech** [comment]
Aem 6.5.11, jdk 11, mac

**Pull request merged**